### PR TITLE
feat: Create the online ingestion role

### PR DESCRIPTION
**Pull request by achals**
Signed-off-by: Achal Shah <achals@gmail.com>

As title, we need a new role that's s used by the online ingestion API. This will further need to be provided to Tecton for  cluster set up.

### CODE REVIEW

**Comment by achals** [comment]
@scottames would really appreciate your eyeballs on this.

File: roles/roles.tf [hunk 1]
```diff
@@ -323,6 +323,154 @@ resource "aws_iam_role_policy_attachment" "common_spark_policy_attachment" {
...
+resource "aws_iam_role_policy_attachment" "lambda_role_dynamodb" {
+  count = var.enable_ingest_api ? 1 : 0
+
+  policy_arn = aws_iam_policy.online_ingestion_role_policy[0].arn
+  role       = aws_iam_role.online_ingestion_role[0].name
+}
+
+// Needed for Lambda to talk to Redis
+// From AWS Docs : Provides minimum permissions for a Lambda function to execute while accessing a resource within a
+// VPC - create, describe, delete network interfaces and write permissions to CloudWatch Logs.
```
Comment (scottames) [inline roles/roles.tf:419]:
```suggestion
/*
Needed for Lambda to talk to Redis
From AWS Docs : Provides minimum permissions for a Lambda function to execute while accessing a resource within a
VPC - create, describe, delete network interfaces and write permissions to CloudWatch Logs.
*/
```

File: roles/roles.tf [hunk 2]
```diff
@@ -323,6 +323,154 @@ resource "aws_iam_role_policy_attachment" "common_spark_policy_attachment" {
...
+resource "aws_iam_role_policy_attachment" "lambda_role_vpc" {
+  count = var.enable_ingest_api ? 1 : 0
+
+  policy_arn = "arn:aws:iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole"
+  role       = aws_iam_role.online_ingestion_role[0].name
+}
+
+#TODO: This should be templated, and the resources it has access to should be constrained.
```
Comment (scottames) [inline roles/roles.tf:427]: This shouldn't be merged until both of these TODOs are resolved.

File: roles/roles.tf [hunk 3]
```diff
@@ -323,6 +323,154 @@ resource "aws_iam_role_policy_attachment" "common_spark_policy_attachment" {
...
+      "elasticloadbalancing:DescribeTags",
+      "elasticloadbalancing:DescribeListeners",
+      "elasticloadbalancing:DescribeRules",
+      "elasticloadbalancing:CreateRule",
+      "elasticloadbalancing:CreateTargetGroup",
+      "elasticloadbalancing:DescribeTargetGroups",
+      "elasticloadbalancing:DescribeTargetHealth",
+      "elasticloadbalancing:RegisterTargets",
+
+      # These permissions are used for orchestrator to discover which stream we publish offline log messages to.
```
Comment (scottames) [inline roles/roles.tf:456]:
```suggestion
      # Used by Tecton orchestration to discover which stream to publish offline log messages to
```

File: roles/roles.tf [hunk 4]
```diff
@@ -323,6 +323,154 @@ resource "aws_iam_role_policy_attachment" "common_spark_policy_attachment" {
...
+    ACCOUNT_ID      = var.account_id
+    DEPLOYMENT_NAME = var.deployment_name
+    REGION          = var.region
+  }
+}
+
+resource "aws_iam_policy" "online_ingestion_role_policy" {
+  count  = var.enable_ingest_api ? 0 : 1
+
+  name   = "tecton-${var.deployment_name}-online-ingestion"
```
Comment (scottames) [inline roles/roles.tf:405]: Can this name be reduced? Many customers will hit the character limit.

File: roles/roles.tf [hunk 5]
```diff
@@ -323,6 +323,154 @@ resource "aws_iam_role_policy_attachment" "common_spark_policy_attachment" {
...
+    ]
+    resources = ["*"]
+  }
+}
+
+# Allow deployment of new versions of the Ingest API and the connection between ALB and Ingest Lambda.
+resource "aws_iam_role_policy" "push_writer_lambda_access" {
+  count = var.enable_ingest_api ? 1 : 0
+
+  name   = "tecton-${var.deployment_name}-online-ingestion-management"
```
Comment (scottames) [inline roles/roles.tf:469]: Can this name be reduced? Many customers will hit the character limit.

File: roles/variables.tf [hunk 6]
```diff
@@ -51,4 +51,10 @@ variable "external_id" {
   default     = ""
   description = "The external id that should be usd by Tecton when assuming the devops role."
   type        = string
+}
+
+variable "enable_ingest_api" {
+  default     = true
+  type        = bool
+  description = "Whether or not to enable resources supporting the Ingest API. Default: true."
 }
```
Comment (scottames) [inline roles/variables.tf:60]: newline

**Review by scottames — CHANGES_REQUESTED** [review]

File: roles/roles.tf [hunk 7]
```diff
@@ -323,6 +323,154 @@ resource "aws_iam_role_policy_attachment" "common_spark_policy_attachment" {
...
+
+      "sqs:SendMessage",
+    ]
+  }
+}
+
+resource "aws_iam_policy" "online_ingestion_logging" {
+  count = var.enable_ingest_api ? 1 : 0
+
+  name   = "tecton-${var.deployment_name}-online-ingestion-logs"
```
[the remark this answers is not in the document]
Comment (achals) [reply roles/roles.tf:356]: I'll try

File: roles/roles.tf [hunk 8]
```diff
@@ -323,6 +323,98 @@ resource "aws_iam_role_policy_attachment" "common_spark_policy_attachment" {
...
+resource "aws_iam_role_policy_attachment" "lambda_role_vpc" {
+  count = var.enable_ingest_api ? 1 : 0
+
+  policy_arn = "arn:aws:iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole"
+  role       = aws_iam_role.online_ingestion_role[0].name
+}
+
+// This file contains the permissions needed by the control plane services to deploy new versions of the Ingest API,
+// update ALB accordingly, and also to discover the offline log on the fly.
+data "template_file" "online_ingestion_management_policy_json" {
```
Comment (poojabhat0) [inline roles/roles.tf:392]: *ingest

**Review by poojabhat0 — DISMISSED** [review]

File: roles/roles.tf (hunk 8 above)
Comment (scottames) [reply roles/roles.tf:392]: Should all of the other `ingestion` be `ingest`?

**Review by scottames — DISMISSED** [review]

**Review by scottames — APPROVED** [review]

**Pull request merged**